modal/_utils/blob_utils.py

```diff
@@ -366,6 +377,113 @@ def source():
+class FileUploadSpec2:
+    source: _FileUploadSource2
+    source_description: Union[str, Path]
+
+    path: str
+    # Raw (unencoded 32 byte) SHA256 sum per 8MiB file block
+    blocks_sha256: list[bytes]
+    mode: int  # file permission bits (last 12 bits of st_mode)
+    size: int
+
+
+    @staticmethod
+    async def from_path(
+        filename: Path,
+        mount_filename: PurePosixPath,
+        mode: Optional[int] = None,
+    ) -> "FileUploadSpec2":
+        # Python appears to give files 0o666 bits on Windows (equal for user, group, and global),
+        # so we mask those out to 0o755 for compatibility with POSIX-based permissions.
+        mode = mode or os.stat(filename).st_mode & (0o7777 if platform.system() != "Windows" else 0o7755)
+
+        def source():
+            return open(filename, "rb")
+
+        return await FileUploadSpec2._create(
+            source,
+            filename,
+            mount_filename,
+            mode,
+        )
+
+
+    @staticmethod
+    async def from_fileobj(
+        source_fp: BinaryIO,
+        mount_filename: PurePosixPath,
+        mode: int
+    ) -> "FileUploadSpec2":
+        def source():
+            new_fd = os.dup(source_fp.fileno())
+            return os.fdopen(new_fd, "rb")
+
+        return await FileUploadSpec2._create(
+            source,
+            str(source),
+            mount_filename,
+            mode,
+        )
+
+
+    @staticmethod
+    async def _create(
+        source: _FileUploadSource2,
+        source_description: Union[str, Path],
+        mount_filename: PurePosixPath,
+        mode: int,
+    ) -> "FileUploadSpec2":
+        # Current position is ignored - we always upload from position 0
+        with source() as source_fp:
+            source_fp.seek(0, os.SEEK_END)
+            size = source_fp.tell()
+
+        blocks_sha256 = await hash_blocks_sha256(source, size)
+
+        return FileUploadSpec2(
+            source=source,
+            source_description=source_description,
+            path=mount_filename.as_posix(),
+            blocks_sha256=blocks_sha256,
+            mode=mode & 0o7777,
+            size=size,
+        )
+
+
+async def hash_blocks_sha256(
+    source: _FileUploadSource2,
+    size: int,
+) -> list[bytes]:
+    def ceildiv(a: int, b: int) -> int:
+        return -(a // -b)
+
+    num_blocks = ceildiv(size, 8 * 1024 * 1024)
+
+    async def hash_block_sha256(block_idx: int) -> bytes:
+        sha256_hash = hashlib.sha256()
+        block_start = block_idx * BLOCK_SIZE
+
+        with source() as block_fp:
+            block_fp.seek(block_start)
+            num_bytes_read = 0
+            while num_bytes_read < BLOCK_SIZE:
+                chunk = block_fp.read(BLOCK_SIZE - num_bytes_read)
+
+                if not chunk:
+                    break
+
+                num_bytes_read += len(chunk)
+                sha256_hash.update(chunk)
+
+        return sha256_hash.digest()
+
+    with asyncio.TaskGroup() as tg:
```

Comment:
Sadly, TaskGroup was added in version 3.11 - all client code needs to support 3.9+ 😬